HACK Week: Load and update site settings

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15371 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR wraps up the notification settings screen by enabling users to configure notification settings for their stores. Changes include:
- Load notification settings for connected sites upon loading the screen.
- Updated the site list to display the notification settings.
- Added the Save button to let users save any changes.
- Error handling and tracking.
- Enabled feature flag for the feature to be available in production.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Pre-requisite: Build and run the app on a physical device because the feature requires notification permission.

### TC1: Loading notification settings
1. Log in to a Jetpack store while authenticated with WPCom.
2. Navigate to Hub Menu > Settings > Notification Settings. Confirm that `settings_notification_settings_tapped` is tracked in Xcode console.
3. Confirm that the screen shows a loading indicator initially and then displays the list of connected stores with their relative notification settings.
4. Tap on any store and confirm that the Notification types screen displays the correct settings states.
5. Navigate back to the Settings screen and disconnect the Internet connection. 
6. Select Notification settings and confirm that the error state is displayed with a retry button after the request times out.

### TC2: Updating notification settings
1. Log in to a Jetpack store while authenticated with WPCom.
2. Navigate to Hub Menu > Settings > Notification Settings. Confirm that `settings_notification_settings_tapped` is tracked in Xcode console.
3. Tap on any store on the store list and toggle any settings then tap Done. Confirm that `notification_settings_update_button_tapped` is tracked in Xcode console.
4. Confirm that the Save button is now enabled. The new settings should also be reflected on the store list.
5. Tap the Save button and confirm that `notification_settings_save_button_tapped` is tracked. 
6. After completion, a notice is displayed indicating the success and the Save button is disabled again. Confirm that `notification_settings_saving_success` is tracked.
7. Repeat step 3 then disconnect the Internet connection.
8. When the request times out, confirm that `notification_settings_saving_failed` is tracked with the error details. An alert should be displayed for the error with a button to retry.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Tested the above test cases on iPhone 16 Pro iOS 18.3.
Unit tests have also been added to confirm the functionalities.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
Store list | Loading | Loading failed | Updating failed | Updating success
--- | ---- | --- | --- | ---
![IMG_7038](https://github.com/user-attachments/assets/02098a50-50e7-47b1-8464-909e096ae786) | ![IMG_7039](https://github.com/user-attachments/assets/509c3d90-4cc7-42b4-9807-aca5155ab4f7) | ![IMG_7045](https://github.com/user-attachments/assets/40646650-cdac-4281-b704-a6d9d50765b5) | ![IMG_7041](https://github.com/user-attachments/assets/1dcf9fe0-a7df-4730-bcc0-a8c0900b7b33) | ![IMG_7043](https://github.com/user-attachments/assets/b5707158-38f1-4598-9ae3-364585e2f2f9)


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Handle loading notification settings for sites

itsmeichigo committed:
Load site settings on the UI

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/240">22.0</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Make NotificationSettings copiable

itsmeichigo committed:
Send settings to site notification view

itsmeichigo committed:
Update bar buttons for notification settings screens

itsmeichigo committed:
Update settings after toggling

itsmeichigo committed:
Handle saving settings

itsmeichigo committed:
Add notice for when saving succeeds

itsmeichigo committed:
Display notification types for each site

itsmeichigo committed:
Add tests for retrieveNotificationSettings

itsmeichigo committed:
Update tests for NotificationSettingsViewModel

itsmeichigo committed:
Add tracking for notification settings

itsmeichigo committed:
Enable feature flag for notification settings

itsmeichigo committed:
Enable Update button only when there are changes

itsmeichigo committed:
Update error handling for notification settings

selanthiraiyan reviewed (approved):
The code looks good to me and tests as described. :shipit: 

---

I found a minor issue while retrying loading notification settings. 

The spinner in the "YOUR STORES" section doesn't show up while retrying from a previous error. 


https://github.com/user-attachments/assets/c3216fb7-9d40-4046-82d9-4281a1e357d4

itsmeichigo committed:
Remove redundant check for notification state and check when to show site list

itsmeichigo committed:
Fix issue displaying progress view when reloading site settings

itsmeichigo commented:
Thank you @selanthiraiyan for the reviews!

> I found a minor issue while retrying loading notification settings. The spinner in the "YOUR STORES" section doesn't show up while retrying from a previous error.

Nice catch! Somehow the progress view would not render when showing the section again, so I replaced it with `ActivityIndicator` in 319ea159b5d31f187d1f5ab4bfa6ee3edbcb2c5e.

I also removed the redundant notification permission check in d9a1e664358a44a892c934a25c1531f660891b12 as it's already fetched upon opening the view, and another check to hide the site list section when tapping Cancel on the site list loading error alert.

Let me know if you have any comments on the new changes, otherwise, I'll merge the PR later before the codefreeze. Thanks!

selanthiraiyan commented:
Thanks for the updates, Huong! The changes still LGTM. 🎉
